workers/workers.go

```diff
@@ -0,0 +1,150 @@
+package workers
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/klippa-app/keda-celery-scaler/celery"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+type celeryWorker struct {
+	Hostname    string
+	LastSeen    time.Time
+	Queues      []string
+	Concurrency int64
+	Active      int64
+}
+
+// A map with information about the current workers.
+var celeryWorkers = map[string]*celeryWorker{}
+
+// A lock for the map above.
+var workerMapLock sync.Mutex
+
+// UpdateWorker will update the current state of a worker with the information
+// from the Celery heartbeat.
+func UpdateWorker(heartbeat celery.Heartbeat) {
+	workerMapLock.Lock()
+	defer workerMapLock.Unlock()
+
+	// Register the worker if we don't know it yet.
+	_, ok := celeryWorkers[heartbeat.Hostname]
+	if !ok {
+		celeryWorkers[heartbeat.Hostname] = &celeryWorker{
+			Hostname: heartbeat.Hostname,
+		}
+	}
+
+	// Set the last seen date and amount of active tasks.
+	celeryWorkers[heartbeat.Hostname].LastSeen = time.Now()
+	celeryWorkers[heartbeat.Hostname].Active = heartbeat.Active
+
+	// If we have the worker info we can automatically fill it in.
+	if heartbeat.XWorkerInfo != nil {
+		celeryWorkers[heartbeat.Hostname].Queues = heartbeat.XWorkerInfo.Queues
+		celeryWorkers[heartbeat.Hostname].Concurrency = heartbeat.XWorkerInfo.Concurrency
+	} else {
+		// No worker info found, map queues and concurrency from hostname.
+		queues, concurrency := getWorkerQueues(heartbeat)
+		celeryWorkers[heartbeat.Hostname].Queues = queues
+		celeryWorkers[heartbeat.Hostname].Concurrency = concurrency
+	}
+
+	// Log the current state of the worker.
+	d, _ := json.Marshal(celeryWorkers[heartbeat.Hostname])
+	log.Debugf("Updated worker %s: %s", heartbeat.Hostname, string(d))
+}
+
+// getWorkerQueues maps the name of the worker to a list of queues set in the
```

Comment:
This functions parses the full map config for every heartbeat. Can this be cached?